### PR TITLE
feat(agg): support `bool_and`, `bool_or`, `jsonb_agg`, `jsonb_object_agg`

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR adds more aggregate functions based on the new macros in #9432.

```
bool_and ( boolean ) → boolean
bool_or ( boolean ) → boolean
jsonb_agg ( anyelement ) → jsonb
jsonb_object_agg ( key "any", value "any" ) → jsonb
```

Limitations:
- The `jsonb_agg` only supports argument of type `boolean`, `int2/4/8`, `float4/8`, `varchar`.
- The `jsonb_object_agg` only supports keys of type `varchar` and values of types shown above.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- SQL commands, functions, and operators

### Release note

Add functions:
```
bool_and ( boolean ) → boolean
bool_or ( boolean ) → boolean
jsonb_agg ( anyelement ) → jsonb
jsonb_object_agg ( key "any", value "any" ) → jsonb
```

Limitations:
- The `jsonb_agg` only supports argument of type `boolean`, `int2/4/8`, `float4/8`, `varchar`.
- The `jsonb_object_agg` only supports keys of type `varchar` and values of types shown above.

</details>

### CODE REVIEW

File: e2e_test/batch/aggregate/jsonb_agg.slt.part [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+query T
+select jsonb_agg(v4) from t;
+----
+[null, null, false, {}]
+
+# query T
+# select jsonb_agg(distinct v1) from t;
+# ----
+# [true, null, false]
```
Comment (richardchien) [inline e2e_test/batch/aggregate/jsonb_agg.slt.part:57]: Why this is commented out?

File: src/expr/src/agg/general.rs [hunk 2]
```diff
@@ -110,6 +110,16 @@ fn count<T>(state: i64, _: T) -> i64 {
...
 
+#[aggregate("bool_and(boolean) -> boolean")]
+fn bool_and(state: bool, input: bool) -> bool {
+    state && input
+}
+
+#[aggregate("bool_or(boolean) -> boolean")]
+fn bool_or(state: bool, input: bool) -> bool {
+    state || input
+}
```
Comment (richardchien) [inline src/expr/src/agg/general.rs:121]: Is it necessary to have e2e tests for this trivial functions? Maybe DocSlt can indeed help.

File: src/expr/src/agg/def.rs [hunk 3]
```diff
@@ -51,17 +51,15 @@ impl AggCall {
         let agg_kind = AggKind::from_protobuf(agg_call.get_type()?)?;
         let args = match &agg_call.get_args()[..] {
             [] => AggArgs::None,
-            [arg] if agg_kind != AggKind::StringAgg => {
-                AggArgs::Unary(DataType::from(arg.get_type()?), arg.get_index() as usize)
-            }
-            [agg_arg, extra_arg] if agg_kind == AggKind::StringAgg => AggArgs::Binary(
+            [arg] => AggArgs::Unary(DataType::from(arg.get_type()?), arg.get_index() as usize),
+            [agg_arg, extra_arg] => AggArgs::Binary(
```
Comment (richardchien) [inline src/expr/src/agg/def.rs:55]: I've also changed this and extract the logic to `AggArgs::from_protobuf` in #9429, may have to resolve conflict for this file.

File: src/frontend/src/optimizer/plan_node/generic/agg.rs [hunk 4]
```diff
@@ -443,7 +451,7 @@ impl<PlanRef: stream::StreamPlanRef> Agg<PlanRef> {
                     }
                 }
                 // TODO: is its state a Table?
-                AggKind::BitAnd | AggKind::BitOr => {
+                AggKind::BitAnd | AggKind::BitOr | AggKind::BoolAnd | AggKind::BoolOr => {
```
Comment (richardchien) [inline src/frontend/src/optimizer/plan_node/generic/agg.rs:454]: Looks like these agg calls with non-append-only input should have `MaterializedInputState`

File: src/frontend/src/optimizer/rule/distinct_agg_rule.rs [hunk 5]
```diff
@@ -281,13 +281,17 @@ impl DistinctAggRule {
...
                     | AggKind::Sum
                     | AggKind::Sum0
                     | AggKind::Avg
                     | AggKind::StringAgg
                     | AggKind::ArrayAgg
+                    | AggKind::JsonbAgg
+                    | AggKind::JsonbObjectAgg
                     | AggKind::FirstValue
                     | AggKind::StddevPop
                     | AggKind::StddevSamp
```
Comment (richardchien) [inline src/frontend/src/optimizer/rule/distinct_agg_rule.rs:297]: Seems the "partial to global" logic is duplicated to `partial_to_total_agg_call` in `generic::Agg` node. Maybe we should extract a common function to ensure consistency later.

**Review by richardchien — APPROVED** [review]
Generally LGTM!

File: src/frontend/src/optimizer/plan_node/generic/agg.rs (hunk 4 above)
Comment (st1page) [reply src/frontend/src/optimizer/plan_node/generic/agg.rs:454]: we can do it better by just storing the count of `1`  in each bit. I am not sure which one is better.
1. rewrite it to multiple `count` in the optimizer, but the pre and post Project needs some bitwise expression and maybe we even need a BIT datatype for better support. 
2. build a new `AggCallState::Table` state for it just like `AggKind::ApproxCountDistinct`
Never mind, I think they(bitAnd and bitOr) are not required now and are less important than the bit_xor(which is often used for verification), so we can design and implement it later.

File: e2e_test/batch/aggregate/jsonb_agg.slt.part (hunk 1 above)
Comment (wangrunji0408) [reply e2e_test/batch/aggregate/jsonb_agg.slt.part]: I thought the order would be unstable but actually it's not.
However, I still have no idea why the result is not `[null, false, true]`. 🤔

File: src/expr/src/agg/general.rs (hunk 2 above)
Comment (wangrunji0408) [reply src/expr/src/agg/general.rs:121]: indeed. thanks for reminding.

**Comment by wangrunji0408** [comment]
It seems that the key order of a jsonb object is unstable.
```json
{"sub-alpha": "GESS90", "t15": "GE1043", "t16": "GE1043"}
{"t15": "GE1043", "t16": "GE1043", "sub-alpha": "GESS90"}
```
so the regress test always failed. 🥵

**Comment by xiangjinwu** [comment]
> It seems that the key order of a jsonb object is unstable.
> 
> ```json
> {"sub-alpha": "GESS90", "t15": "GE1043", "t16": "GE1043"}
> {"t15": "GE1043", "t16": "GE1043", "sub-alpha": "GESS90"}
> ```

https://github.com/risingwavelabs/risingwave/blob/7d49b88c72089f27bb2ba111a62d6b2f8af9f0e7/e2e_test/batch/types/jsonb.slt.part#L141-L152

`serde_json::Value` (without `preserve_order` feature) uses lexical order (`str as Ord`).

**Comment by wangrunji0408** [comment]
So shall we keep the regression test disabled for `jsonb_object_agg`?

**Comment by xiangjinwu** [comment]
Yes. We keep the regress test disabled (because our output is different from PostgreSQL), and optionally makes a copy of the test with modified expectation in e2e.

**Pull request merged**